cache/memcached_client.go

```diff
@@ -0,0 +1,695 @@
+
+	// MaxGetMultiBatchSize specifies the maximum number of keys a single underlying
+	// GetMulti() should run. If more keys are specified, internally keys are splitted
+	// into multiple batches and fetched concurrently, honoring MaxGetMultiConcurrency parallelism.
+	// If set to 0, the max batch size is unlimited.
+	MaxGetMultiBatchSize int `yaml:"max_get_multi_batch_size"`
+
+	// DNSProviderUpdateInterval specifies the DNS discovery update interval.
+	DNSProviderUpdateInterval time.Duration `yaml:"dns_provider_update_interval"`
+
+	// AutoDiscovery configures memached client to perform auto-discovery instead of DNS resolution
+	AutoDiscovery bool `yaml:"auto_discovery"`
+}
+
+func (c *MemcachedClientConfig) validate() error {
+	if len(c.Addresses) == 0 {
+		return errMemcachedConfigNoAddrs
+	}
+
+	// Avoid panic in time ticker.
+	if c.DNSProviderUpdateInterval <= 0 {
+		return errMemcachedDNSUpdateIntervalNotPositive
+	}
+
+	// Set async only available when MaxAsyncConcurrency > 0.
+	if c.MaxAsyncConcurrency <= 0 {
+		return errMemcachedMaxAsyncConcurrencyNotPositive
+	}
+
+	return nil
+}
+
+// parseMemcachedClientConfig unmarshals a buffer into a MemcachedClientConfig with default values.
+func parseMemcachedClientConfig(conf []byte) (MemcachedClientConfig, error) {
+	config := defaultMemcachedClientConfig
+	if err := yaml.Unmarshal(conf, &config); err != nil {
+		return MemcachedClientConfig{}, err
+	}
+
+	return config, nil
+}
+
+type memcachedClient struct {
+	logger   log.Logger
+	config   MemcachedClientConfig
+	client   memcachedClientBackend
+	selector updatableServerSelector
+
+	// Name provides an identifier for the instantiated Client
+	name string
+
+	// Address provider used to keep the memcached servers list updated.
+	addressProvider AddressProvider
+
+	// Channel used to notify internal goroutines when they should quit.
+	stop chan struct{}
+
+	// Channel used to enqueue async operations.
+	asyncQueue chan func()
+
+	// Gate used to enforce the max number of concurrent GetMulti() operations.
+	getMultiGate gate.Gate
+
+	// Wait group used to wait all workers on stopping.
+	workers sync.WaitGroup
+
+	// Tracked metrics.
+	clientInfo prometheus.GaugeFunc
+	operations *prometheus.CounterVec
+	failures   *prometheus.CounterVec
+	skipped    *prometheus.CounterVec
+	duration   *prometheus.HistogramVec
+	dataSize   *prometheus.HistogramVec
+}
+
+// AddressProvider performs node address resolution given a list of clusters.
+type AddressProvider interface {
+	// Resolves the provided list of memcached cluster to the actual nodes
+	Resolve(context.Context, []string) error
+
+	// Returns the nodes
+	Addresses() []string
+}
+
+type memcachedGetMultiResult struct {
+	items map[string]*memcache.Item
+	err   error
+}
+
+// NewMemcachedClient makes a new RemoteCacheClient.
+func NewMemcachedClient(logger log.Logger, name string, conf []byte, reg prometheus.Registerer) (RemoteCacheClient, error) {
```

Comment:
Delete this function. Was inherited from Thanos, but we don't want to pass the config as a marshalled `[]byte` slice.